### PR TITLE
Pre-2022-03-31 warning/debashification pass

### PR DESCRIPTION
**Pull request by nabijaczleweli**
* **What does this PR do?**
Few things, see individual commit messages.
Primarily, this is a warning pass (GCC 11.2.0-14 on x32 and Clang 15.0.0-++20220302071730+0e38b295435b-1\~exp1\~20220302071825.172 on amd64) for my distconfig:
```
./configure --mandir=/usr/share/man --libexecdir=/usr/libexec --with-mailpath=/var/mail --gpgme --lua --notmuch --gnutls --gss --idn --mixmaster --sasl --tokyocabinet --lmdb --zlib --zstd --testing \
  CFLAGS='-g -O3 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-sign-compare' \
  CXXFLAGS='-g -O3 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-sign-compare'
```
`-Wno-sign-compare` because there's way too many, the other two because they're benign.
A few remain.

This:
```
cc -g -O3 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-sign-compare -std=c99 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include -I/usr/include/lua5.4 -DNCURSES_WIDECHAR -isystem /usr/include/mit-krb5 -I. -I. -Wall  -I./test -I./fuzz -MT conn/dlg_verifycert.o -MD -MP -MF conn/dlg_verifycert.Tpo -c -o conn/dlg_verifycert.o conn/dlg_verifycert.c
conn/sasl.c: In function ‘mutt_sasl_start’:
conn/sasl.c:291:23: warning: cast between incompatible function types from ‘int (*)(void *, int,  const char *)’ to ‘int (*)(void)’ [-Wcast-function-type]
  291 |   callbacks[0].proc = (int (*)(void)) mutt_sasl_cb_log;
      |                       ^
conn/sasl.c: In function ‘mutt_sasl_get_callbacks’:
conn/sasl.c:396:20: warning: cast between incompatible function types from ‘int (*)(void *, int,  const char **, unsigned int *)’ to ‘int (*)(void)’ [-Wcast-function-type]
  396 |   callback->proc = (int (*)(void)) mutt_sasl_cb_authname;
      |                    ^
conn/sasl.c:401:20: warning: cast between incompatible function types from ‘int (*)(void *, int,  const char **, unsigned int *)’ to ‘int (*)(void)’ [-Wcast-function-type]
  401 |   callback->proc = (int (*)(void)) mutt_sasl_cb_authname;
      |                    ^
conn/sasl.c:406:20: warning: cast between incompatible function types from ‘int (*)(sasl_conn_t *, void *, int,  sasl_secret_t **)’ {aka ‘int (*)(struct sasl_conn *, void *, int,  struct sasl_secret **)’} to ‘int (*)(void)’ [-Wcast-function-type]
  406 |   callback->proc = (int (*)(void)) mutt_sasl_cb_pass;
      |                    ^
```
is unavoidable but I still dislike it. A union would've cost them nothing.

This:
```
cc -g -O3 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-sign-compare -std=c99 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include -I/usr/include/lua5.4 -DNCURSES_WIDECHAR -isystem /usr/include/mit-krb5 -I. -I. -Wall  -I./test -I./fuzz -MT mutt/slist.o -MD -MP -MF mutt/slist.Tpo -c -o mutt/slist.o mutt/slist.c
mutt/regex.c: In function ‘mutt_replacelist_apply’:
mutt/regex.c:421:15: warning: ‘strncpy’ output may be truncated copying between 0 and 1023 bytes from a string of length 2047 [-Wstringop-truncation]
  421 |               strncpy(&dst[tlen], src, cpysize);
      |               ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
is checked in the line above I think, but.

This:
```
cc -g -O3 -Wall -Wextra -Wno-unused-parameter -Wno-missing-field-initializers -Wno-sign-compare -std=c99 -D_ALL_SOURCE=1 -D_GNU_SOURCE=1 -D__EXTENSIONS__ -I/usr/include -I/usr/include/lua5.4 -DNCURSES_WIDECHAR -isystem /usr/include/mit-krb5 -I. -I. -Wall  -I./test -I./fuzz -MT notmuch/notmuch.o -MD -MP -MF notmuch/notmuch.Tpo -c -o notmuch/notmuch.o notmuch/notmuch.c
notmuch/notmuch.c: In function ‘rename_filename’:
notmuch/notmuch.c:1240:29: warning: ‘%s’ directive output may be truncated writing 3 bytes into a region of size between 0 and 4095 [-Wformat-truncation=]
 1240 |   snprintf(buf, buflen, "%s/%s/%s%s", folder,
      |                             ^~
notmuch/notmuch.c:1240:3: note: ‘snprintf’ output between 6 and 12291 bytes into a destination of size 4096
 1240 |   snprintf(buf, buflen, "%s/%s/%s%s", folder,
      |   ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 1241 |            (e->read || e->old) ? "cur" : "new", filename, suffix);
      |            ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
is actually a potential problem, since AFAICT it's entirely possible for this to not get null-terminated, the next line is a `strcmp()`.

As long as I'm complaining: if you `./configure` with `-Werror` literally nothing gets picked up. This was fun to debug.

Tests pass for me on both platforms. Cursory glance at running is Fine, will continue to do so until the release.

### CODE REVIEW

**Comment by nabijaczleweli** [comment]
Also: 
```
Checking libs for stringprep_check_version...-lidn
Checking for idna_to_unicode_utf8_from_utf8...not found
Checking for idna_to_unicode_8z8z...ok
Checking for idna_to_ascii_from_utf8...not found
Checking for idna_to_ascii_8z...ok
Checking for idna_to_ascii_lz...ok
Checking for idna_to_ascii_from_locale...not found
Checking for ...ok
```
at least its all okay :)

File: mutt/file.h [hunk 1]
```diff
@@ -39,19 +40,6 @@ typedef uint8_t ReadLineFlags;             ///< Flags for mutt_file_read_line(),
...
-/**
- * struct timespec - Time value with nanosecond precision
- */
-struct timespec
-{
-  time_t tv_sec; ///< Number of seconds since the epoch
-  long tv_nsec;  ///< Number of nanosecond, on top
-};
-#endif
-
```
Comment (gahr) [inline mutt/file.h:54]: `HAVE_STRUCT_TIMESPEC` is defined by our `configure` script.

**Comment by gahr** [comment]
LGTM, except for the change in mutt/file.h

File: mutt/file.h (hunk 1 above)
Comment (nabijaczleweli) [reply mutt/file.h]: Oh, so it is! I think I got this from my initial test with `-Werror` where it didn't get picked up. Dropped.

**Review by gahr — APPROVED** [review]

**Comment by flatcap** [comment]
Thanks for the fixes, @nabijaczleweli!

A few notes...

I can't update your branch.
Did you untick the box that says, `[X] Allow edits by maintainers` ?

I've shortened some of the log messages.
It's good practice to try to limit them to ~50 chars or so.

If you're running the tests, it's a good idea to also enable [ASAN (Address Sanitizer)](https://neomutt.org/dev/analysis/asan).

```sh
./configure ... --testing --asan
```

Perhaps we could replace all the `strncpy()`s with our `mutt_str_copy()`.
It's guaranteed to NUL-terminate the string.

---

These sort of fixes are always welcome.
Thanks

**Comment by nabijaczleweli** [comment]
Just blindly mass-replacing all possible `strncpy()`s with `mutt_str_copy()`ies got me a result worse than nothing, so it probably wants some finesse, which I'm sadly not able to provide for most of the affected code.

**Comment by flatcap** [comment]
np, thanks for looking

**Pull request merged**